examples/customresources/demos/complexmodules/moduletest/module_test.go

```diff
@@ -0,0 +1,171 @@
+package main_test
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"testing"
+	"time"
+
+	"go.viam.com/test"
+	goutils "go.viam.com/utils"
+	"go.viam.com/utils/pexec"
+	"go.viam.com/utils/rpc"
+
+	"go.viam.com/rdk/config"
+	"go.viam.com/rdk/examples/customresources/apis/gizmoapi"
+	"go.viam.com/rdk/examples/customresources/apis/summationapi"
+	"go.viam.com/rdk/logging"
+	"go.viam.com/rdk/robot"
+	"go.viam.com/rdk/robot/client"
+	"go.viam.com/rdk/testutils"
+	"go.viam.com/rdk/testutils/robottestutils"
+	"go.viam.com/rdk/utils"
+)
+
+func TestComplexModules(t *testing.T) {
+	logger, observer := logging.NewObservedTestLogger(t)
+
+	var port int
+	success := false
+	for portTryNum := 0; portTryNum < 10; portTryNum++ {
+		// Modify the example config to run directly, without compiling the module first.
+		cfgFilename, portLocal, err := modifyCfg(t, utils.ResolveFile("examples/customresources/demos/complexmodules/module.json"), logger)
+		port = portLocal
+		test.That(t, err, test.ShouldBeNil)
+
+		serverPath := testutils.BuildTempModule(t, "web/cmd/server/")
+
+		// start the viam server with a temporary home directory so that it doesn't collide with
+		// the user's real viam home directory
+		testTempHome := t.TempDir()
+		server := pexec.NewManagedProcess(pexec.ProcessConfig{
+			Name:        serverPath,
+			Args:        []string{"-config", cfgFilename},
+			CWD:         utils.ResolveFile("./"),
+			Environment: map[string]string{"HOME": testTempHome},
+			Log:         true,
+		}, logger.AsZap())
+
+		err = server.Start(context.Background())
+		test.That(t, err, test.ShouldBeNil)
+
+		if robottestutils.WaitForServing(observer, port) {
+			success = true
+			defer func() {
+				test.That(t, server.Stop(), test.ShouldBeNil)
+			}()
+			break
+		} else {
+			server.Stop()
+		}
+	}
+	test.That(t, success, test.ShouldBeTrue)
+
+	rc, err := connect(port, logger)
+	test.That(t, err, test.ShouldBeNil)
+	defer func() {
+		test.That(t, rc.Close(context.Background()), test.ShouldBeNil)
+	}()
+
+	// Gizmo is a custom component model and API.
+	t.Run("Test Gizmo", func(t *testing.T) {
+		res, err := rc.ResourceByName(gizmoapi.Named("gizmo1"))
+		test.That(t, err, test.ShouldBeNil)
+
+		giz := res.(gizmoapi.Gizmo)
+		ret1, err := giz.DoOne(context.Background(), "1.0")
+		test.That(t, err, test.ShouldBeNil)
+		test.That(t, ret1, test.ShouldBeTrue)
+
+		ret2, err := giz.DoOneClientStream(context.Background(), []string{"1.0", "2.0", "3.0"})
+		test.That(t, err, test.ShouldBeNil)
+		test.That(t, ret2, test.ShouldBeTrue)
+
+		ret3, err := giz.DoOneServerStream(context.Background(), "1.0")
+		test.That(t, err, test.ShouldBeNil)
+		test.That(t, ret3, test.ShouldResemble, []bool{true, false, true, false})
+
+		ret3, err = giz.DoOneBiDiStream(context.Background(), []string{"1.0", "2.0", "3.0"})
+		test.That(t, err, test.ShouldBeNil)
+		test.That(t, ret3, test.ShouldResemble, []bool{true, true})
```

Comment:
we drop the first message in client streaming and server streaming, but this seems to be an existing issue (see complexmodule tests), will file a ticket and fix later